### ISSUE TITLE
`tests/test_leak.py::test_markup_leaks` fails with JIT-enabled Python 3.13.0b1

### ISSUE THREAD

**Issue by mgorny** [pallets/markupsafe#445]
When Python 3.13.0b1 is built with `--enable-experimental-jit`, I'm getting the following test failure:

```pytb
$ tox -e py313
.pkg-cpython313: _optional_hooks> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg-cpython313: get_requires_for_build_wheel> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg-cpython313: build_wheel> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py313: install_package> python -I -m pip install --force-reinstall --no-deps /tmp/markupsafe/.tox/.tmp/package/2/MarkupSafe-3.0.0.dev0-cp313-cp313-linux_x86_64.whl
py313: commands[0]> pytest -v --tb=short --basetemp=/tmp/markupsafe/.tox/py313/tmp
========================================================= test session starts =========================================================
platform linux -- Python 3.13.0b1, pytest-8.2.0, pluggy-1.5.0 -- /tmp/markupsafe/.tox/py313/bin/python
cachedir: .tox/py313/.pytest_cache
rootdir: /tmp/markupsafe
configfile: pyproject.toml
testpaths: tests
collected 74 items                                                                                                                    

tests/test_escape.py::test_escape[markupsafe._native--] PASSED                                                                  [  1%]
tests/test_escape.py::test_escape[markupsafe._native-abcd&><'"efgh-abcd&amp;&gt;&lt;&#39;&#34;efgh] PASSED                      [  2%]
tests/test_escape.py::test_escape[markupsafe._native-&><'"efgh-&amp;&gt;&lt;&#39;&#34;efgh] PASSED                              [  4%]
tests/test_escape.py::test_escape[markupsafe._native-abcd&><'"-abcd&amp;&gt;&lt;&#39;&#34;] PASSED                              [  5%]
tests/test_escape.py::test_escape[markupsafe._native-\u3053\u3093\u306b\u3061\u306f&><'"\u3053\u3093\u3070\u3093\u306f-\u3053\u3093\u306b\u3061\u306f&amp;&gt;&lt;&#39;&#34;\u3053\u3093\u3070\u3093\u306f] PASSED [  6%]
tests/test_escape.py::test_escape[markupsafe._native-&><'"\u3053\u3093\u3070\u3093\u306f-&amp;&gt;&lt;&#39;&#34;\u3053\u3093\u3070\u3093\u306f] PASSED [  8%]
tests/test_escape.py::test_escape[markupsafe._native-\u3053\u3093\u306b\u3061\u306f&><'"-\u3053\u3093\u306b\u3061\u306f&amp;&gt;&lt;&#39;&#34;] PASSED [  9%]
tests/test_escape.py::test_escape[markupsafe._native-\U0001f363\U0001f362&><'"\U0001f37a xyz-\U0001f363\U0001f362&amp;&gt;&lt;&#39;&#34;\U0001f37a xyz] PASSED [ 10%]
tests/test_escape.py::test_escape[markupsafe._native-&><'"\U0001f37a xyz-&amp;&gt;&lt;&#39;&#34;\U0001f37a xyz] PASSED          [ 12%]
tests/test_escape.py::test_escape[markupsafe._native-\U0001f363\U0001f362&><'"-\U0001f363\U0001f362&amp;&gt;&lt;&#39;&#34;] PASSED [ 13%]
tests/test_exception_custom_html.py::test_exception_custom_html[markupsafe._native] PASSED                                      [ 14%]
tests/test_leak.py::test_markup_leaks[markupsafe._native] FAILED                                                                [ 16%]
tests/test_markupsafe.py::test_adding[markupsafe._native] PASSED                                                                [ 17%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._native-<em>%s</em>-<bad user>-<em>&lt;bad user&gt;</em>] PASSED [ 18%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._native-<em>%(username)s</em>-data1-<em>&lt;bad user&gt;</em>] PASSED [ 20%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._native-%i-3.14-3] PASSED                                        [ 21%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._native-%.2f-3.14-3.14] PASSED                                   [ 22%]
tests/test_markupsafe.py::test_type_behavior[markupsafe._native] PASSED                                                         [ 24%]
tests/test_markupsafe.py::test_html_interop[markupsafe._native] PASSED                                                          [ 25%]
tests/test_markupsafe.py::test_missing_interpol[markupsafe._native-foo] PASSED                                                  [ 27%]
tests/test_markupsafe.py::test_missing_interpol[markupsafe._native-42] PASSED                                                   [ 28%]
tests/test_markupsafe.py::test_missing_interpol[markupsafe._native-args2] PASSED                                                [ 29%]
tests/test_markupsafe.py::test_tuple_interpol[markupsafe._native] PASSED                                                        [ 31%]
tests/test_markupsafe.py::test_dict_interpol[markupsafe._native] PASSED                                                         [ 32%]
tests/test_markupsafe.py::test_escaping[markupsafe._native] PASSED                                                              [ 33%]
tests/test_markupsafe.py::test_unescape[markupsafe._native] PASSED                                                              [ 35%]
tests/test_markupsafe.py::test_format[markupsafe._native] PASSED                                                                [ 36%]
tests/test_markupsafe.py::test_format_map[markupsafe._native] PASSED                                                            [ 37%]
tests/test_markupsafe.py::test_formatting_empty[markupsafe._native] PASSED                                                      [ 39%]
tests/test_markupsafe.py::test_custom_formatting[markupsafe._native] PASSED                                                     [ 40%]
tests/test_markupsafe.py::test_complex_custom_formatting[markupsafe._native] PASSED                                             [ 41%]
tests/test_markupsafe.py::test_formatting_with_objects[markupsafe._native] PASSED                                               [ 43%]
tests/test_markupsafe.py::test_escape_silent[markupsafe._native] PASSED                                                         [ 44%]
tests/test_markupsafe.py::test_splitting[markupsafe._native] PASSED                                                             [ 45%]
tests/test_markupsafe.py::test_mul[markupsafe._native] PASSED                                                                   [ 47%]
tests/test_markupsafe.py::test_escape_return_type[markupsafe._native] PASSED                                                    [ 48%]
tests/test_markupsafe.py::test_soft_str[markupsafe._native] PASSED                                                              [ 50%]
tests/test_escape.py::test_escape[markupsafe._speedups--] PASSED                                                                [ 51%]
tests/test_escape.py::test_escape[markupsafe._speedups-abcd&><'"efgh-abcd&amp;&gt;&lt;&#39;&#34;efgh] PASSED                    [ 52%]
tests/test_escape.py::test_escape[markupsafe._speedups-&><'"efgh-&amp;&gt;&lt;&#39;&#34;efgh] PASSED                            [ 54%]
tests/test_escape.py::test_escape[markupsafe._speedups-abcd&><'"-abcd&amp;&gt;&lt;&#39;&#34;] PASSED                            [ 55%]
tests/test_escape.py::test_escape[markupsafe._speedups-\u3053\u3093\u306b\u3061\u306f&><'"\u3053\u3093\u3070\u3093\u306f-\u3053\u3093\u306b\u3061\u306f&amp;&gt;&lt;&#39;&#34;\u3053\u3093\u3070\u3093\u306f] PASSED [ 56%]
tests/test_escape.py::test_escape[markupsafe._speedups-&><'"\u3053\u3093\u3070\u3093\u306f-&amp;&gt;&lt;&#39;&#34;\u3053\u3093\u3070\u3093\u306f] PASSED [ 58%]
tests/test_escape.py::test_escape[markupsafe._speedups-\u3053\u3093\u306b\u3061\u306f&><'"-\u3053\u3093\u306b\u3061\u306f&amp;&gt;&lt;&#39;&#34;] PASSED [ 59%]
tests/test_escape.py::test_escape[markupsafe._speedups-\U0001f363\U0001f362&><'"\U0001f37a xyz-\U0001f363\U0001f362&amp;&gt;&lt;&#39;&#34;\U0001f37a xyz] PASSED [ 60%]
tests/test_escape.py::test_escape[markupsafe._speedups-&><'"\U0001f37a xyz-&amp;&gt;&lt;&#39;&#34;\U0001f37a xyz] PASSED        [ 62%]
tests/test_escape.py::test_escape[markupsafe._speedups-\U0001f363\U0001f362&><'"-\U0001f363\U0001f362&amp;&gt;&lt;&#39;&#34;] PASSED [ 63%]
tests/test_exception_custom_html.py::test_exception_custom_html[markupsafe._speedups] PASSED                                    [ 64%]
tests/test_leak.py::test_markup_leaks[markupsafe._speedups] FAILED                                                              [ 66%]
tests/test_markupsafe.py::test_adding[markupsafe._speedups] PASSED                                                              [ 67%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._speedups-<em>%s</em>-<bad user>-<em>&lt;bad user&gt;</em>] PASSED [ 68%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._speedups-<em>%(username)s</em>-data1-<em>&lt;bad user&gt;</em>] PASSED [ 70%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._speedups-%i-3.14-3] PASSED                                      [ 71%]
tests/test_markupsafe.py::test_string_interpolation[markupsafe._speedups-%.2f-3.14-3.14] PASSED                                 [ 72%]
tests/test_markupsafe.py::test_type_behavior[markupsafe._speedups] PASSED                                                       [ 74%]
tests/test_markupsafe.py::test_html_interop[markupsafe._speedups] PASSED                                                        [ 75%]
tests/test_markupsafe.py::test_missing_interpol[markupsafe._speedups-foo] PASSED                                                [ 77%]
tests/test_markupsafe.py::test_missing_interpol[markupsafe._speedups-42] PASSED                                                 [ 78%]
tests/test_markupsafe.py::test_missing_interpol[markupsafe._speedups-args2] PASSED                                              [ 79%]
tests/test_markupsafe.py::test_tuple_interpol[markupsafe._speedups] PASSED                                                      [ 81%]
tests/test_markupsafe.py::test_dict_interpol[markupsafe._speedups] PASSED                                                       [ 82%]
tests/test_markupsafe.py::test_escaping[markupsafe._speedups] PASSED                                                            [ 83%]
tests/test_markupsafe.py::test_unescape[markupsafe._speedups] PASSED                                                            [ 85%]
tests/test_markupsafe.py::test_format[markupsafe._speedups] PASSED                                                              [ 86%]
tests/test_markupsafe.py::test_format_map[markupsafe._speedups] PASSED                                                          [ 87%]
tests/test_markupsafe.py::test_formatting_empty[markupsafe._speedups] PASSED                                                    [ 89%]
tests/test_markupsafe.py::test_custom_formatting[markupsafe._speedups] PASSED                                                   [ 90%]
tests/test_markupsafe.py::test_complex_custom_formatting[markupsafe._speedups] PASSED                                           [ 91%]
tests/test_markupsafe.py::test_formatting_with_objects[markupsafe._speedups] PASSED                                             [ 93%]
tests/test_markupsafe.py::test_escape_silent[markupsafe._speedups] PASSED                                                       [ 94%]
tests/test_markupsafe.py::test_splitting[markupsafe._speedups] PASSED                                                           [ 95%]
tests/test_markupsafe.py::test_mul[markupsafe._speedups] PASSED                                                                 [ 97%]
tests/test_markupsafe.py::test_escape_return_type[markupsafe._speedups] PASSED                                                  [ 98%]
tests/test_markupsafe.py::test_soft_str[markupsafe._speedups] PASSED                                                            [100%]

============================================================== FAILURES ===============================================================
________________________________________________ test_markup_leaks[markupsafe._native] ________________________________________________
tests/test_leak.py:31: in test_markup_leaks
    assert len(counts) == 1
E   assert 2 == 1
E    +  where 2 = len({31213, 31214})
_______________________________________________ test_markup_leaks[markupsafe._speedups] _______________________________________________
tests/test_leak.py:31: in test_markup_leaks
    assert len(counts) == 1
E   assert 2 == 1
E    +  where 2 = len({31954, 31955})
======================================================= short test summary info =======================================================
FAILED tests/test_leak.py::test_markup_leaks[markupsafe._native] - assert 2 == 1
FAILED tests/test_leak.py::test_markup_leaks[markupsafe._speedups] - assert 2 == 1
==================================================== 2 failed, 72 passed in 0.35s =====================================================
py313: exit 1 (0.61 seconds) /tmp/markupsafe> pytest -v --tb=short --basetemp=/tmp/markupsafe/.tox/py313/tmp pid=842994
  py313: FAIL code 1 (2.36=setup[1.75]+cmd[0.61] seconds)
  evaluation failed :( (2.78 seconds)
```

Curiously, enabling the `gc.collect()` call doesn't seem to help here.

Environment:

- Gentoo Linux amd64
- Python version: 3.13.0b1
- MarkupSafe version: 52d35f57ee0ba9dd8d74032f7741c27f27b05edf

**Comment by davidism** [comment]
Appears to be passing on main. I'm on Mac and did the following:

```
$ ./configure --disable-gil --with-openssl="$(brew --prefix openssl)" && make -j
```

```
$ virtualenv -p ~/Python-3.13.0b1/python.exe .venv313
$ . .venv313/bin/activate.fish
$ pip install -r requirements/dev.txt && pip install -e .
$ PYTHON_GIL=0 pytest
```

However, if I take out the `if pypy` condition, and always call `gc.collect()`, then it _does_ fail on every cpython version (not just 3.13).

Frankly, I do not understand this test (and I did not write it, and there's no comments explaining it). Why do we get a different count after GC, except on PyPy where we get a different count without GC? Why do we only get 2 different counts over 20x1000 loops? Why are we treating this second count as a memory leak in MarkupSafe, it seems like it's just the behavior of Python ref counts and GC at work. I would expect a memory leak to result in a different count _most_ times, not only 2 different counts.

**Comment by mgorny** [comment]
> Appears to be passing on main. I'm on Mac and did the following:
> 
> ```
> $ ./configure --disable-gil --with-openssl="$(brew --prefix openssl)" && make -j
> ```

You've missed `--enable-experimental-jit`, though I don't know if it's support on Mac.

**Comment by davidism** [comment]
Ah, I missed that it was the JIT and not the GIL. Anyway, my confusion over the entire test still stands.

**Comment by davidism** [comment]
JIT doesn't build on Mac yet.

**Comment by davidism** [comment]
Moving `gc.collect()` to before the loops rather than during them works for cpython and PyPy, but not 3.13 JIT. I think I'm just going to say `assert count < 3`. I'm guessing some internals stabilize over time resulting in different counts. If this test becomes flaky again in the future I might just remove it.

**Comment by davidism** [comment]
Closed by #446

**Comment by mgorny** [comment]
Thanks! I can confirm that they pass for me now.
